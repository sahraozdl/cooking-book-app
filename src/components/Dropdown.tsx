"use client";

import {
  Listbox,
  ListboxButton,
  ListboxOption,
  ListboxOptions,
  Field,
  Label,
} from "@headlessui/react";
import { Fragment } from "react";

export type Option = { id: string; name: string };
type DropdownProps<T extends boolean = false> = {
  label: string;
  options: Option[];
  selected: T extends true ? Option[] : Option | null;
  setSelected: (val: T extends true ? Option[] : Option | null) => void;
  name: string;
  multiple?: T;
};

export default function Dropdown<T extends boolean = false>({
  label,
  options,
  selected,
  setSelected,
  multiple = false as T,
}: DropdownProps<T>) {
  const isSelected = (option: Option) => {
    if (multiple && Array.isArray(selected)) {
      return selected.some((item) => item.id === option.id);
    }
    return (selected as Option)?.id === option.id;
  };

  const removeItem = (id: string) => {
    if (multiple && Array.isArray(selected)) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      setSelected(selected.filter((item) => item.id !== id) as any);
    }
  };

  return (
    <Field className="flex flex-col my-4">
      <Label className="mb-1 text-sm font-medium">{label}</Label>
      <Listbox value={selected} onChange={setSelected} multiple={multiple}>
        <div className="relative">
          <ListboxButton className="border p-2 rounded bg-black w-full text-left text-white min-h-[42px] flex flex-wrap gap-2">
            {multiple && Array.isArray(selected) && selected.length > 0 ? (
              selected.map((item) => (
                <span
                  key={item.id}
                  className="bg-gray-800 px-2 py-1 rounded text-sm flex items-center gap-3"
                >
                  {item.name}
                  <span
                    role="button"
                    tabIndex={0}
                    className="text-red-400 hover:text-red-600 cursor-pointer text-2xl select-none"
                    onClick={(e) => {
                      e.stopPropagation();
                      removeItem(item.id);
                    }}
                    onKeyDown={(e) => {
                      if (e.key === "Enter" || e.key === " ") {
                        e.preventDefault();
                        removeItem(item.id);
                      }
                    }}
                    aria-label={`Remove ${item.name}`}
                  >
                    Ã—
                  </span>
                </span>
              ))
            ) : (
              <span>
                {!selected ? `Select ${label}` : (selected as Option).name}
              </span>
            )}
          </ListboxButton>

          <ListboxOptions className="absolute z-10 mt-1 w-full border bg-black rounded shadow-lg max-h-40 overflow-auto text-white">
            {options.map((option) =>  (
                <ListboxOption key={option.id} value={option} as={Fragment}>
                  {({ active }) => (
                    <li
                      className={`cursor-pointer px-4 py-2 ${
                        isSelected(option) ? "bg-blue-900 font-semibold" : ""
                      } ${active ? "bg-gray-900" : ""}`}
                    >
                      {option.name}
                    </li>
                  )}
                </ListboxOption>
              ))
            }
          </ListboxOptions>
        </div>
      </Listbox>

      
    </Field>
  );
}
