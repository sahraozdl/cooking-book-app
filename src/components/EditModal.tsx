import React, { ReactNode } from "react";
import { Dialog,DialogPanel,DialogTitle } from "@headlessui/react";

interface EditModalProps {
  isOpen: boolean;
  onClose: () => void;
  title?: string;
  children: ReactNode;
}

export default function EditModal({ isOpen, onClose, title = "Edit", children }: EditModalProps) {
  return (
    <Dialog open={isOpen} onClose={onClose} className="fixed z-50 inset-0">
      <div className="fixed inset-0 bg-black bg-opacity-10" aria-hidden="true" />
      <div className="fixed inset-0 flex items-center justify-center overflow-y-scroll">
        <DialogPanel className="bg-gray-800 p-6 rounded-xl max-w-lg w-full shadow-xl">
          <DialogTitle className="text-lg font-bold mb-4">{title}</DialogTitle>
          {children}
        </DialogPanel>
      </div>
    </Dialog>
  );
}
