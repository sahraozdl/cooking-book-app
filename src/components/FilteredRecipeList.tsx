"use client";

import { useEffect, useState } from "react";
import { useSearchParams } from "next/navigation";
import { getFilteredSortedRecipesFromParams } from "@/app/actions/getFilteredSortedRecipesFromParams";
import { RecipeWithID } from "@/types";
import RecipeList from "./RecipeList";

export default function FilteredRecipeList() {
  const searchParams = useSearchParams();
  const [recipes, setRecipes] = useState<RecipeWithID[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchRecipes = async () => {
      setLoading(true);
      const data = await getFilteredSortedRecipesFromParams(searchParams);
      setRecipes(data);
      setLoading(false);
    };

    fetchRecipes();
  }, [searchParams]);

  return (
    <RecipeList
      recipes={recipes}
      loading={loading}
    />
  );
}
