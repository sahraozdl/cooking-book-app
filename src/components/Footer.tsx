export const Footer = () => {
  return (
    <footer className="bg-orange-300 p-4">
        <p className="text-center text-red-950">&copy; {new Date().getFullYear()} Cooking Book App. All rights reserved.</p>
    </footer>
  );
}