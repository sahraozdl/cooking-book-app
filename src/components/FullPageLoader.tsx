'use client';

import { FireIcon } from '@phosphor-icons/react';
import React from 'react';

const FullPageLoader = () => {
  return (
    <div className="fixed inset-0 z-50 flex flex-col items-center justify-center bg-white">
      <FireIcon size={64} weight="fill" className="w-16 h-16 text-orange-500 animate-bounce" />
      <p className="mt-4 text-lg text-gray-700">Firing up the kitchen...</p>
    </div>
  );
};

export default FullPageLoader;
