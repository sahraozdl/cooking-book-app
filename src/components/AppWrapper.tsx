'use client';

import { Navbar } from '@/components/Navbar';
import { Footer } from '@/components/Footer';
import { useUser } from '@/store/UserContext';
import FullPageLoader from '@/components/FullPageLoader';

const AppWrapper = ({ children }: { children: React.ReactNode }) => {
  const { loading } = useUser();

  if (loading) return <FullPageLoader />;

  return (
    <div className="flex flex-col min-h-screen">
      <header className="sticky top-0 z-50 bg-white shadow-sm">
        <Navbar />
      </header>
      <main className="flex-grow px-4 py-6 bg-orange-50">{children}</main>
      <Footer />
    </div>
  );
};

export default AppWrapper;
