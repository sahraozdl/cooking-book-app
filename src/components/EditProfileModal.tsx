"use client";

import { useState } from "react";
import { updateProfile } from "@/store/firebase/firestoreUser";
import { UserTypes } from "@/types/recipes";
import { useUser } from "@/components/UserContext";

export default function EditProfileModal({
  isOpen,
  onClose,
  user,
}: {
  isOpen: boolean;
  onClose: () => void;
  user: UserTypes;
}) {
  const [name, setName] = useState(user.name ?? "");

  const { setUser } = useUser();

  const handleUpdate = async () => {
    if (!user.id) throw new Error("User ID is required");
    await updateProfile(user.id, { name });
    setUser((prev) => (prev ? { ...prev, name } : prev));
    onClose();
  };

  if (!isOpen) return null;

  return (
    <div className="fixed inset-0 bg-black/50 flex justify-center items-center z-50">
      <div className="bg-black p-6 rounded shadow-xl w-full max-w-md">
        <h2 className="text-xl font-semibold mb-4">Edit Profile</h2>

        <label className="block mb-2">
          Name
          <input
            className="w-full border p-2 rounded"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </label>

        <div className="flex justify-end space-x-2 mt-4">
          <button onClick={onClose} className="btn-secondary">
            Cancel
          </button>
          <button onClick={handleUpdate} className="btn">
            Save
          </button>
        </div>
      </div>
    </div>
  );
}
