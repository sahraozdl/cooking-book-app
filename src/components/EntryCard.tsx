"use client";

import Image from "next/image";
import Link from "next/link";
import { useState, useEffect } from "react";
import {
  toggleRecipeLike,
  toggleRecipeSave,
} from "@/app/actions/firestoreRecipes";
import { useUser } from "@/components/UserContext"; // adjust based on your setup
import { RecipeWithID } from "@/types/recipes";
import { useRouter } from "next/navigation";

interface EntryCardProps {
  entry: RecipeWithID;
  showAuthor?: boolean;
}

export default function EntryCard({
  entry,
  showAuthor = true,
}: EntryCardProps) {
  const { user } = useUser(); // assumes you have a context with user.id
  const [liked, setLiked] = useState(false);
  const [saved, setSaved] = useState(false);

  useEffect(() => {
    if (!user) return;
    setLiked(user.id ? entry.likedBy?.includes(user.id) ?? false : false);
    setSaved(user.id ? entry.savedBy?.includes(user.id) ?? false : false);
  }, [entry, user]);

  const router = useRouter();

  const handleToggle = async (type: "like" | "save") => {
    if (!user?.id) {
      router.push("/login");
      return;
    }
    if (!entry.id) return;

    const isLike = type === "like";
    const toggler = isLike ? toggleRecipeLike : toggleRecipeSave;

    try {
      const newState = isLike ? !liked : !saved;

      // Optimistically update UI
      if (isLike) {
        setLiked(newState);
      } else {
        setSaved(newState);
      }

      // Update count and array locally
      if (isLike) {
        entry.likeCount = Math.max(
          (entry.likeCount ?? entry.likedBy?.length ?? 0) + (newState ? 1 : -1),
          0
        );
      } else {
        entry.saveCount = Math.max(
          (entry.saveCount ?? entry.savedBy?.length ?? 0) + (newState ? 1 : -1),
          0
        );
      }

      if (newState) {
        entry[`${type}dBy`] = [...(entry[`${type}dBy`] ?? []), user.id];
      } else {
        entry[`${type}dBy`] =
          entry[`${type}dBy`]?.filter((id) => id !== user.id) ?? [];
      }

      await toggler(entry.id, user.id);
    } catch (error) {
      console.error(`Failed to toggle ${type}:`, error);
    }
  };

  return (
    <div className="border rounded-xl p-4 dark:border-gray-700 shadow-sm bg-white dark:bg-gray-900">
      <div className="flex items-start gap-4">
        {entry.strMealThumb && (
          <Image
            src={entry.strMealThumb}
            alt={entry.strMeal}
            width={100}
            height={100}
            className="rounded-lg object-cover w-24 h-24"
          />
        )}

        <div className="flex-1 space-y-1">
          <h3 className="text-lg font-semibold dark:text-white">
            {entry.strMeal}
          </h3>
          <p className="text-sm text-gray-600 dark:text-gray-400 line-clamp-3">
            {entry.strInstructions}
          </p>

          <div className="text-xs text-gray-500 dark:text-gray-400 mt-1 space-x-2">
            <span>🍽️ Servings: {entry.servingsId}</span>
            <span>🌶️ Difficulty: {entry.difficultyId}</span>
            <span>🌍 Cuisine: {entry.cuisineId}</span>
          </div>

          {showAuthor && entry.authorId && (
            <p className="text-sm">
              By: {entry.isAnonymous ? "Anonymous" : entry.authorName}
            </p>
          )}

          <div className="flex gap-4 mt-2">
            <button
              onClick={() => handleToggle("like")}
              className={`text-sm ${
                liked ? "text-red-500" : "text-gray-500"
              } hover:underline`}
            >
              ❤️ {liked ? "Liked" : "Like"} (
              {entry.likeCount ?? entry.likedBy?.length ?? 0})
            </button>

            <button
              onClick={() => handleToggle("save")}
              className={`text-sm ${
                saved ? "text-green-600" : "text-gray-500"
              } hover:underline`}
            >
              📌 {saved ? "Saved" : "Save"} (
              {entry.saveCount ?? entry.savedBy?.length ?? 0})
            </button>

            <Link
              href={`/recipes/${entry.id}`}
              className="ml-auto text-purple-600 hover:underline text-sm"
            >
              View full recipe
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
}
