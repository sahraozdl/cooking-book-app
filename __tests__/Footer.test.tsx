import { render, screen } from '@testing-library/react';
import { Footer } from '@/components/Footer';

describe('Footer', () => {
  it('renders copyright text', () => {
    render(<Footer />);
    expect(screen.getByText(/Cooking Book App/i)).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Footer />);
    expect(asFragment()).toMatchSnapshot();
  });
});
