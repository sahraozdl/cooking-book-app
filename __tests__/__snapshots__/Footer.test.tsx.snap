// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Footer matches snapshot 1`] = `
<DocumentFragment>
  <footer
    class="bg-orange-300 p-4"
  >
    <p
      class="text-center text-red-950"
    >
      Â© 2025 Cooking Book App. All rights reserved.
    </p>
  </footer>
</DocumentFragment>
`;
